### ISSUE TITLE
Serenity/JS 3.0 - Test fails immediately and does not await `Wait` timeout if `Text` of a `PageElement` in a `List` can't be retrieved

### ISSUE THREAD

**Issue by viper3400** [serenity-js/serenity-js#1339]
Based on [this test](https://github.com/serenity-js/serenity-js/blob/3480ba44d83ff663a6a7c3e51ab1f977776cc9d0/integration/web-specs/spec/screenplay/models/PageElements.patterns.spec.ts#L118) I wrote the following one. 

It will fail with the `LogicError: Can't retrieve the first item from a list with 0 items: [ ]` (as expected).

But it fails immediately, not as expected, after the `Wait` timeout, in this case 30 seconds:

```ts
it(`waits for x seconds if the text of an element is not present before fail`, () =>
    actorCalled('Peggy').attemptsTo(
        Wait.upTo(Duration.ofSeconds(30)).until(
            Text.of(
                parents()
                .where(Text.ofAll(children()), contain('tea'))
                .where(Text.ofAll(children()), contain('coffee'))
                .where(Text.ofAll(children()), contain('juice'))
                // there's no parent container with all the three items
                .first()
            ), includes('tea')
        ))
);
```

In my case, there are lists, I have to wait until some list get the expected content. As workaround I would have to work with explicit fixed wait timeouts, what is not very convenient. 

see https://serenityjs-serenityjs-adtizt5mmv3.ws-eu71.gitpod.io/

**Comment by viper3400** [comment]
@jan-molak Any ideas here?

**Comment by jan-molak** [comment]
We could have a look at the errors thrown by List when the collection is empty. Maybe Wait could ignore errors caused in those circumstances since they might be recoverable

**Comment by jan-molak** [comment]
@viper3400 so I think what we could do to make the behaviour more intuitive is to replace the `LogicError` with some `List`-specific error, like `ListItemNotFound` or something like that here in `get()`:

https://github.com/serenity-js/serenity-js/blob/3480ba44d83ff663a6a7c3e51ab1f977776cc9d0/packages/core/src/screenplay/questions/List.ts#L83-L93

`first()`:

https://github.com/serenity-js/serenity-js/blob/3480ba44d83ff663a6a7c3e51ab1f977776cc9d0/packages/core/src/screenplay/questions/List.ts#L59-L69

and `last()`:

https://github.com/serenity-js/serenity-js/blob/3480ba44d83ff663a6a7c3e51ab1f977776cc9d0/packages/core/src/screenplay/questions/List.ts#L71-L81

Then change the error handling logic in `Wait` so that when such `ListItemNotFound` error is detected, we restart the polling:

https://github.com/serenity-js/serenity-js/blob/3480ba44d83ff663a6a7c3e51ab1f977776cc9d0/packages/core/src/screenplay/interactions/Wait.ts#L281-L297

**Comment by manuelfidalgo** [comment]
Adding also some comments from gitter chat:
Looks clear that, in order to use that expression, you need 2 conditions to be met:
- element list is not empty
- there is an element in the list with certain text.

I can think of 3 main scenarios where you want to use that:
1. **Direct interaction** (for example: Click.on(PageElements..where...))
At least with wdio retry mechanism, it should keep retrying x times the same as with a simpler expression and, as a serenity user, I don't care if PageElements does not contain elements, just keep retrying, because I expect the element to be there at certain point. I cannot think of any case where you want an early exit there if there are no elements.

2. **Used in combination with Wait.until**.
Is the same as above: the thing is that, if there are no elements, normally is because they have not been loaded yet, so you need to keep trying until both conditions are met
I think you and @viper3400 described well the problem and the possible solution.

3. **Used with not**(isVisible(PageElements..where...)) && Wait.until (not(isVisible(PageElements..where...)))
in that case, if the list is empty, the condition (nand? xD) is met so the flow should continue normally

**Comment by viper3400** [comment]
@manuelfidalgo , @jan-molak, wow, guys, not sure how to answer here 😅.

I tried to follow up you conversation at Gitter. Still, I have the impression we're probably mixing things up here (or mixing to much into it).

I don't wanted to question the current `Wait` implementation in this issue.  I wanted to point out an incompatibility in special use cases when filtered lists are involved.

I think with the `Wait` interaction Serentiy/JS has chosen [the explicit wait approach](https://webdriver.io/docs/autowait/) not the implicit one, right? 

Then, @manuelfidalgo , not sure, what you mean with wdio retry mechanism in your first point, or how this should apply?

Do you mean https://webdriver.io/docs/retry? I do not use this, but it seems, it just enables us to re-run a failed step, or a failed spec file? It would not retry the specific action, when a specific selector was not found. A click on a selector, is just one _part_ of a test step, isn't it?

- enter username into page element located by selector1
- enter the password into page element located by selector2
- click save element located by selector3

The first selector that can not be found will fail the test step. But the wdio retry will include the whole test step, not just the retry of the failed selector, right?

Otherwise, do you refer to the https://webdriver.io/docs/clioptions and the `--waitforTimeout` option (timeout for all waitForXXX commands)? I think this is pretty much the same Serenity/JS does now, just using it's own `Wait` implementation.

I suggest to use concrete use cases to discuss our points to avoid misunderstandings.

Use case A) Given a shopping list feature with a preloaded list of items
- As a user I ensure that all expected items are visible
- As a user I click on (and select) a product from the preloaded list
```
<html>
  <product-list>
   <item>milk</item>
   <item>sugar</item>
   <item>butter</item>
  </product-list>
</html>
```

Remark: This is a custom implementation that is not the [HTML option tag](https://www.w3schools.com/tags/tag_option.asp)

First question:
- How to build the selectors for this list?

```ts
const productList = () => 
  PageElement
    .located(By.css(product-list))
    .describedAs('product list');

const shoppingListItems = () =>
   PageElements
     .located(By.css(item))
     .of(productList)
     .describedAs('all shopping list items');

const shoppingListItemWithName = (productName: string) =>
  shoppingListItems
    .where(Text, equals(text))
    .first();
```
Then our test would look somehow like:

```ts
// 1
Ensure.that(shoppingListItemWithName('milk'), isVisible());
Ensure.that(shoppingListItemWithName('sugar'), isVisible());
Ensure.that(shoppingListItemWithName('butter'), isVisible());
Ensure.that(shoppingListItems().count, equals(3));
Ensure.that(shoppingListItemWithName('olives'), not(isVisible()));

// 2
Click.on(shoppingListItemWithName('milk'));
```

I would expect all this interactions to fail immediately if an element was not found, or the condition is not met. What is the reason, I should retry this? Elements are expected to be on the page, given, that I expect the page loading has finished at the time I do the checks.


Lets extend our use case:

Use Case B) Given a shopping list feature with a dropdown
- As a user I can click on the dropdown
- The system fills this dropdown dynamically (takes some time ...)
- As a user I expect the list to be fully loaded after a certain amount of time
- As a user I ensure that all expected items are visible
- As a user I click on (and select) a product from the preloaded list

before user clicks:
```html
<html>
  <product-list class="dropdown" onClick="loadShoppingListItems()">
  </product-list>
</html>
```

after user clicks:
```html
<html>
  <product-list class="dropdown" onClick="loadShoppingListItems()">
   <item>milk</item>
   <item>sugar</item>
   <item>butter</item>
  </product-list>
</html>
```

```ts
// 1
Wait.until(shoppingListItems().count, equals(3));
Ensure.that(shoppingListItemWithName('milk'), isVisible());
Ensure.that(shoppingListItemWithName('sugar'), isVisible());
Ensure.that(shoppingListItemWithName('butter'), isVisible());
Ensure.that(shoppingListItemWithName('olives'), not(isVisible()));

// 2
Click.on(shoppingListItemWithName('milk'));
```

Now my approach was not to wait until the list count has the expected value (what is a valid workaround for the issue I try do describe), but to `Wait` for the element to appear in the list directly:

```ts
// 1
Wait.until(shoppingListItemWithName('milk'), isVisible());
Wait.until(shoppingListItemWithName('sugar'), isVisible());
Wait.until(shoppingListItemWithName('butter'), isVisible());

// 2
Click.on(shoppingListItemWithName('milk'));
```

This fails immediately, tough I would expect it to fail after the explicit wait timeout. This is due to the implementation [mentioned by Jan](https://github.com/serenity-js/serenity-js/issues/1339#issuecomment-1284106102).

As I said, the workaround with the `count` is valid, but in my real life scenario I have lists the are probably filled up by other tests in parallel and I can't say what the current count of list items is at a given time (and I really don't care about the count, I just make sure, that my items under test are visible or not).

Let's have another use case to cover the `not(isVisible())` thing a bit better, by allow our user to delete an item from the list:

Use Case C) Given a shopping list feature with a dropdown
- As a user I can click on the dropdown
- The system fills this dropdown dynamically
- As a user I expect the list to be fully loaded after a certain amount of time
- As a user I ensure that all expected items are visible
- As a user I can _delete_ an item _in this list_
- The system removes the item from the list (takes some time ...)
- As a user I expect the item to be not visible after a certain amount of time


```ts
// 1
Wait.until(shoppingListItems().count, equals(3));
Ensure.that(shoppingListItemWithName('milk'), isVisible());
Ensure.that(shoppingListItemWithName('sugar'), isVisible());
Ensure.that(shoppingListItemWithName('butter'), isVisible());
Ensure.that(shoppingListItemWithName('olives'), not(isVisible()));

// 2
Click.on(DeleteShoppingListItemWithName('milk'));

// 3
Wait.until(shoppingListItems().count, equals(2));
Ensure.that(shoppingListItemWithName('olives'), not(isVisible()));
```

In step 3) I could have try to do a `Wait.until(shoppingListItemWithName('olives'), not(isVisible()));` but as we learned, this will _not wait_ and fail immediately, because the list is still empty.

Could you explain your points using some examples? Is it the same issue as mine?

**Comment by viper3400** [comment]
@jan-molak, concerning to our former discussion and the necessary changes in 
https://github.com/serenity-js/serenity-js/blob/3480ba44d83ff663a6a7c3e51ab1f977776cc9d0/packages/core/src/screenplay/interactions/Wait.ts#L338-L341

I added a `pollIfPollingActive()` sub-function into `waitUntil` that encapsulate some functionality. It is called twice. Once at the place where it was before and again when the `LogicError` occurs, so we have a recursion here, until the `TimeoutExpiredError` is thrown.

```ts
function waitUntil(expectation: () => Promise<boolean> | boolean, pollingInterval: Duration): { start: () => Promise<void>, stop: () => void } {
    let delay: { start: () => Promise<void>, stop: () => void };
    let pollingActive = false;    

    async function pollIfPollingActive(): Promise<void> {
        if (pollingActive) {
            delay = waitFor(pollingInterval);

            await delay.start();

            await poll();
        }

    }
    async function poll(): Promise<void> {
        try {
            const expectationIsMet = await expectation();

            if (expectationIsMet) {
                delay?.stop();
                return;
            }

            await pollIfPollingActive();
        } catch (error) {
            delay?.stop();
            if (error instanceof LogicError) {                       
                await pollIfPollingActive();
            } else {
                throw error;
            }
        }
        
    }

    return {
        async start () {
            pollingActive = true;
            await poll();
        },

        stop () {
            delay?.stop();
            pollingActive = false;
        }
    };
}
```

If this is still the way to go I could try to implement the `ListItemNotFound` error, create some valuable tests and create a PR. But yeah, just makes sense for me to invest more time here, if this would find it's way to the main branch and can be useful for erverybody.

**Comment by jan-molak** [comment]
@viper3400 - yes, this looks good, thanks for working on this!

I think `ListItemNotFoundError` will need to extend `RuntimeError` and you could add it to [`core/io/errors`](https://github.com/serenity-js/serenity-js/tree/main/packages/core/src/errors)

**Comment by manuelfidalgo** [comment]
Hi @viper3400,
sorry for the confusion and thanks for clarifying the issue with use cases, I think it is important to think all the different scenarios because the behavior is related.
When I talk about wdio retry mechanism, I talk about [this](https://webdriver.io/docs/options/#connectionretrycount), because I expect that wdio will retry some requests, avoiding some traditional "pure" Selenium problems.
My case might be different than yours but, when I use something like your case **b)** I normally don't check that all the options are loaded, I focus on selecting option "x". I don't like to `Wait.until (list().count, equals(3))` because '3' is a magic number and for me (in most cases) is irrelevant. What is relevant is **certain text to be present**
When I build a `Wait.until()`, I would expect it to finish if ALL conditions are met (success) xor timeout has been reached (fail): if list.count()==0 I would expect to keep retrying because, eventually, the list will be populated.
Not sure if this is your idea and also @jan-molak , but anyway it is not a big deal, the good thing is that I can create a generic Task that checks both things (list is not empty and Text exists)

**Comment by viper3400** [comment]
@jan-molak I tried to fix the issue:

- Created the `ListItemNotFoundError`: https://github.com/viper3400/serenity-js/blob/fix/wait-for-empty-list/packages/core/src/errors/ListItemNotFoundError.ts
- Applied it to the `List`: https://github.com/viper3400/serenity-js/blob/fix/wait-for-empty-list/packages/core/src/screenplay/questions/List.ts
- Update the `List` tests: https://github.com/viper3400/serenity-js/blob/fix/wait-for-empty-list/packages/core/spec/screenplay/questions/List.spec.ts
- Tried to fix the issue in `Wait.until`: https://github.com/viper3400/serenity-js/blob/fix/wait-for-empty-list/packages/core/src/screenplay/interactions/Wait.ts
- Tried to write tests in the `Wait` spec: https://github.com/viper3400/serenity-js/blob/fix/wait-for-empty-list/packages/core/spec/screenplay/interactions/Wait.spec.ts

It now fails like expected after the timeout has expired when the list is empty: https://github.com/viper3400/serenity-js/blob/a120ab8b06d6d7c77608457f5fe3a458bebcf93a/packages/core/spec/screenplay/interactions/Wait.spec.ts#L109-L127

But I'd like to test the situation when after a delay the list gets filled. I tried a lot, to manipulate the content of the list after a javascript timeout, but the list object seems to be instantiated at the beginning and can't be changed:

https://github.com/viper3400/serenity-js/blob/a120ab8b06d6d7c77608457f5fe3a458bebcf93a/packages/core/spec/screenplay/interactions/Wait.spec.ts#L129-L149

How can I simulate this?

**Comment by jan-molak** [comment]
Hey @viper3400! I'm on holiday at the moment, so will try to suggest something once I find someplace with a decent Internet connection

**Comment by viper3400** [comment]
Then @jan-molak you should enjoy your holidays, decent internet connections will be back soon enough! 😏 Hope you have a great time.

**Comment by jan-molak** [comment]
Looks great so far!

> But I'd like to test the situation when after a delay the list gets filled

Yes, this sounds like a good test.

The trouble you encountered [here](https://github.com/viper3400/serenity-js/blob/a120ab8b06d6d7c77608457f5fe3a458bebcf93a/packages/core/spec/screenplay/interactions/Wait.spec.ts#L129-L149) is cause by the fact that `item` is a value that gets evaluated and doesn't get re-evaluated again:
```
Wait.upTo(Duration.ofSeconds(7)).until(item, isGreaterThan(0)), // `item` is a value
```
If instead, you used a `Question`, it would get re-evaluated on every interval:
```
const myList = () => Question.about('my list', actor => [ 1, 2, 3 ]))
```
then:
```
Wait.upTo(Duration.ofSeconds(7)).until(myList(), isGreaterThan(0))
```

Now, you can do interesting things with such question, for example, you could make the answer dependent on the `StopWatch`:

```
const myList = () =>
  Question.about('my list', actor => {
    const elapsedTime = UseAStopwatch.as(actor).stopwatch.elapsedTime.inMilliseconds();
    
    if (elapsedTime < ...) {
      return []; 
    }
    
    return [ 1, 2, 3 ];
  }))
```

Does it help?

**Comment by viper3400** [comment]
Yes, thanks. I will try to implement it  asap and come back with the PR. :+1:

**Comment by viper3400** [comment]
Hi, tried to implement it, but did not manage it.

First, `stopwatch.elapsedTime` is not a public member of `UseAStopwatch`. Next, `myList` returns not a `List` (what would be the subject under test) but an array of numbers.

What I tried so far,

I exposed `elapsed` from the local `Stopwatch` class:

```ts
    elapsed() {
        return this.elapsedTime;
    }
```

I used this with you example but returned a `List`:
```ts
const myList = () =>
                Question.about('my list', actor => {
                    const elapsedTime = UseAStopwatch.as(actor).stopwatch.elapsed().inMilliseconds();

                    if (elapsedTime < 1000) {
                        return List.of<number>([]);
                    }

                    return List.of<number>([ 1, 2, 3 ]);
                });
```

I copied `equals` from assertions:

```ts
function equals<Expected>(expectedValue: Answerable<Expected>): Expectation<Expected> {
    return Expectation.thatActualShould<Expected, Expected>('equal', expectedValue)
        .soThat((actual, expected) => equal(actual, expected));
}
```

And then I tried to let an actor attempt to:

```ts
Wait.upTo(Duration.ofSeconds(7)).until(myList().first(), equals(1))
```

But now types are incompatible:
```
Argument of type 'Expectation<number>' is not assignable to parameter of type 'Expectation<Question<Promise<number>> & Interaction & { isPresent(): Question<Promise<boolean>>; } & QuestionAdapterFieldDecorator<number>>'.
  Type 'number' is not assignable to type 'Question<Promise<number>> & Interaction & { isPresent(): Question<Promise<boolean>>; } & QuestionAdapterFieldDecorator<number>'.
    Type 'number' is not assignable to type 'Question<Promise<number>>'.ts(2345)
```

How could I ensure that the first item of my list is what I expect?

**Comment by jan-molak** [comment]
Hey @viper3400 - your approach seems sensible.
Could you please raise a PR with the code throwing this error, perhaps I could have a look what's going on?

**Comment by viper3400** [comment]
@jan-molak raised the draft PR. It's probably just a type mismatch in my assertion, that TS won't compile at all. But then I'm not sure if the rest would work out, either. 😬

**Comment by jan-molak** [comment]
Hey @viper3400, please have a look at #1355; your PR needed just a slight tweak 👍🏻

**Comment by viper3400** [comment]
Solved with merge of #1355 and release of 3.0.0-rc.34. Thanks, @jan-molak
